### PR TITLE
Decouple database storage and file source APIs

### PR DESCRIPTION
**Pull request by pozdnyakov**
This patches provides the following improvements:
- decouples `FileSource` implementation and the database control logic 
- introduces separate interfaces for ambient cache and offline regions storage
- introduces `FileSourceManager::getDatabaseStorage()` API returning pointer to `JointDatabaseStorage`, thus we obviate downcasting of `FileSource` to `DatabaseFileSource`

### CODE REVIEW

File: bin/offline.cpp [hunk 1]
```diff
@@ -163,29 +163,28 @@ int main(int argc, char *argv[]) {
 
 
     util::RunLoop loop;
-    std::shared_ptr<DatabaseFileSource> fileSource = std::static_pointer_cast<DatabaseFileSource>(
-        std::shared_ptr<FileSource>(FileSourceManager::get()->getFileSource(
-            FileSourceType::Database,
-            ResourceOptions().withAccessToken(token).withBaseURL(apiBaseURL).withCachePath(output))));
+    std::shared_ptr<JointDatabaseStorage> databaseStorage = FileSourceManager::get()->getDatabaseStorage(
```
Comment (tmpsantos) [inline bin/offline.cpp]: Maybe a better name... SQLiteDatabaseStorage?

Comment (chloekraw) [reply bin/offline.cpp]: Not so sure about adding SQLite to the API name. Are we moving to a non-SQLite implementation and doing this in order to maintain backwards compatibility?

What's problematic about the existing name?

**Comment by chloekraw** [comment]
@pozdnyakov 👋 hey, what's currently blocking this?

**Comment by pozdnyakov** [comment]
> @pozdnyakov wave hey, what's currently blocking this?

I just need to find a time slot to finalize this :slightly_smiling_face:

**Pull request closed without merging**